### DISCUSSION TITLE
[python/webgpu] faster indexing along multiple dimensions (w/o unnecessary copies)

### DISCUSSION THREAD

**Discussion by sluijs** [microsoft/onnxruntime#23217] [Performance Q&A]
I'm working with large tensors of rank 5 (e.g. w/ shape `[20, 60, 512, 512, 1]`). In the model outlined below, I'm simply trying to index the tensor along multiple dimensions. Executing this in PyTorch-CPU and onnxruntime in Python takes around ~10-30 microseconds. However, executing w/ onnxruntime-web and the WebGPU execution provider this takes ~120 ms using pre-allocated GPU buffers.

```python
class Model(nn.Module):
    def forward(self, x: torch.Tensor, index: torch.Tensor) -> torch.Tensor:
        return x[index[0], index[1], :, :, index[4]]
```

The following graph seems to indicate that ONNX first indexes the first dim `x[index[0]]` and sequentially indexes the other dims. If I understand correctly, each step would create a copy of the underlying data. Is there a way to prevent these copies from happening, or only a copy of the final slice?

<img width="420" alt="afbeelding" src="https://github.com/user-attachments/assets/3180fd5e-2f78-4f8e-9e80-e0d38c1fb1db" />
